pkg/didcomm/crypto/jwe/authcrypt/encrypt.go

```diff
@@ -0,0 +1,392 @@
+	var encodedRecipients []Recipient
+	for _, e := range c.recipients {
+		rec, err := c.encodeRecipient(sharedSymKey, e)
+		if err != nil {
+			return nil, err
+		}
+		encodedRecipients = append(encodedRecipients, *rec)
+	}
+	return encodedRecipients, nil
+}
+
+// encodeRecipient will encode the sharedKey (cek) with recipientKey
+// by generating a new ephemeral key to be used by the recipient to decrypt the cek
+func (c *Crypter) encodeRecipient(sharedSymKey *[chacha20poly1305.KeySize]byte, recipientKey *[chacha20poly1305.KeySize]byte) (*Recipient, error) {
+	// generate a random APU value (Agreement PartyUInfo: https://tools.ietf.org/html/rfc7518#section-4.6.1.2)
+	apu := make([]byte, 64)
+	_, err := RandReader.Read(apu)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a new ephemeral key for the recipient and return its APU
+	kek, err := c.generateRecipientCek(apu, recipientKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a new (chacha20poly1035) cipher with this new key to encrypt the shared key (cek)
+	cipher, err := createCipher(c.nonceSize, kek)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a new nonce
+	nonce := make([]byte, c.nonceSize)
+	_, err = RandReader.Read(nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	// encrypt symmetric shared key cek using the recipient's ephemeral key
+	kekOutput := cipher.Seal(nil, nonce, sharedSymKey[:], nil)
+
+	// kekOutput has a length of len(sharedSymKey.priv) + poly1035.TagSize
+	// fetch the tag from the tail of kekOutput
+	tag := kekOutput[len(kekOutput)-poly1305.TagSize:]
+	// now fetch the cipherText from the head of kekOutput
+	sharedKeyCipher := kekOutput[0 : len(kekOutput)-poly1305.TagSize]
+
+	return c.buildRecipient(sharedKeyCipher, apu, nonce, tag, recipientKey)
+}
+
+// buildRecipient will build a proper JSON formatted Recipient
+func (c *Crypter) buildRecipient(symKeyCipher []byte, apu []byte, nonce []byte, tag []byte, recipientKey *[chacha20poly1305.KeySize]byte) (*Recipient, error) {
+	oid, err := encryptOID(recipientKey, []byte(base58.Encode(c.sender.pub[:])))
+	if err != nil {
+		return nil, err
+	}
+
+	recipientHeaders := RecipientHeaders{
+		APU: base64.URLEncoding.EncodeToString(apu),
+		IV:  base64.URLEncoding.EncodeToString(nonce),
+		Tag: base64.URLEncoding.EncodeToString(tag),
+		KID: base58.Encode(recipientKey[:]),
+		OID: base64.URLEncoding.EncodeToString(oid),
+	}
+
+	recipient := &Recipient{
+		EncryptedKey: base64.URLEncoding.EncodeToString(symKeyCipher),
+		Header:       recipientHeaders,
+	}
+
+	return recipient, nil
+}
+
+// generateRecipientCek will generate an ephemeral symmetric key for the recipientKey to
+// be used for encrypting the cek.
+// it will return this new key along with the corresponding APU or an error if it fails.
+func (c *Crypter) generateRecipientCek(apu []byte, recipientKey *[chacha20poly1305.KeySize]byte) ([]byte, error) {
```

Comment:
generateRecipientCEK

Reply:
This function will need further review and more tests. It is hard to follow. Assume this is coming in a followup.

Reply:
yes, this is following the php example instructions, interop testing is required to ensure this works with libraries in other packages.

I will rename the function to CEK

Reply:
done